protocols/mockserver/mockserver.go

```diff
@@ -0,0 +1,268 @@
+// Copyright 2017 HyperHQ Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// gRPC mock server
+
+package mockserver
+
+import (
+	"errors"
+
+	google_protobuf "github.com/golang/protobuf/ptypes/empty"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	pb "github.com/kata-project/agent/protocols/grpc"
+)
+
+type pod struct {
+	containers map[string]*container
+}
+
+type container struct {
+	id   string
+	init string // init process name
+	proc map[string]*process
+}
+
+type process struct {
+	id   string
+	proc *pb.Process
+}
+
+type mockServer struct {
+	pod *pod
+}
+
+func NewMockServer() *grpc.Server {
+	mock := &mockServer{}
+	serv := grpc.NewServer()
+	pb.RegisterHyperstartServiceServer(serv, mock)
+
+	return serv
+}
+
+func validateOCISpec(spec *pb.Spec) error {
+	if spec == nil || spec.Process == nil {
+		return errors.New("invalid container spec")
+	}
+	return nil
+}
+
+func (m *mockServer) CreateContainer(ctx context.Context, req *pb.CreateContainerRequest) (*google_protobuf.Empty, error) {
+	if m.pod == nil {
+		return nil, errors.New("pod not created")
+	}
+	if req.ContainerId == "" {
+		return nil, errors.New("container ID must be set")
+	}
+	if m.pod.containers[req.ContainerId] != nil {
+		return nil, errors.New("container ID already taken")
+	}
+	if err := validateOCISpec(req.OCI); err != nil {
+		return nil, err
+	}
+
+	c := &container{
+		id:   req.ContainerId,
+		proc: make(map[string]*process),
+	}
+	c.init = "init"
+	p := &process{
+		id:   c.init,
+		proc: req.OCI.Process,
+	}
+	c.proc[c.init] = p
+	m.pod.containers[req.ContainerId] = c
+
+	return &google_protobuf.Empty{}, nil
+}
+
+func (m *mockServer) StartContainer(ctx context.Context, req *pb.StartContainerRequest) (*google_protobuf.Empty, error) {
+	if m.pod == nil {
```

Comment:
These checks occur in lots of different functions. You might be able to refactor to add a `func (m *mockServer) checkValues()` or similar to avoid the code duplication?